### ISSUE TITLE
VST freezes on Arch

### ISSUE THREAD

**Issue by mateuszmnm** [LMMS/lmms#1875]
## Steps to reproduce:
* start LMMS
* drag VeSTige into main screen
* click on VeSTige's green folder icon
* select a .DLL VST plugin

### Expected:
* loading plugin window appears
* plugin loads

### Actual:
* loading plugin window appears
* one of the CPU cores goes 100%
* LMMS stops responding
* it stays like that forever

Screenshot: http://i.imgur.com/2f5YURI.png
Affected system: Arch 64 bit
LMMS versions: 1.1.1, 1.1.2, 1.1.3
Wine versions used: at least 4 different - 1.6.4. 1.7.22, 1.7.33, 1.7.38
Example VST: Synth1
More details: https://lmms.io/forum/viewtopic.php?f=7&t=1749

Some strace dump: https://gist.github.com/obszczymucha/161b9fb25ce0a48b6af3 (it goes like that forever)

**Comment by obszczymucha** [comment]
Doh, posted above on wrong account...

**Comment by obszczymucha** [comment]
I've profiled LMMS with Valgrind, here's the result:
https://www.dropbox.com/s/46p9g4bnknvs272/callgrind.out.2681?dl=0

I still don't know what to make out of this, but maybe someone else beats me to it.

**Comment by obszczymucha** [comment]
If that helps, I've managed to debug LMMS into a point where execution stops at: https://github.com/LMMS/lmms/blob/stable-1.1/plugins/vst_base/VstPlugin.cpp#L180

which is ``waitForInitDone()``

I'm guessing LMMS waits for a signal from remote plugin and either doesn't get it or gets it but doesn't process it properly (or locked?).

Additionally I can see that wineserver process starts by doing 'ps aux'.

**Comment by badosu** [comment]
Great bug report @obszczymucha! :+1:

**Comment by obszczymucha** [comment]
Bad news. We never get inside this loop: https://github.com/LMMS/lmms/blob/stable-1.1/include/RemotePlugin.h#L973

**Comment by tresf** [comment]
Possible duplicate of #1567.
Related: #1097 #1736

Last time this surfaced, it wasn't a problem with our code.  @badosu do you run Arch?  I'll try setting up a build environment now per the Arch instructions on our wiki.

**Comment by tresf** [comment]
>  I'll try setting up LMMS now per the Arch instructions on our download page.

Scrap that... Arch is a manual install.  I'll have to defer this unless someone has an Arch 2015 VirtualBox image handy.

**Comment by badosu** [comment]
Reproduced on my machine as well, can't figure out why this happens.

**Comment by tresf** [comment]
@badosu does this happen with other VSTs (i.e. DSK musicbox?)

**Comment by badosu** [comment]
Yep, tested on lots of other VSTs. This is not a particular VST problem but with the messagery between wine and lmms as indicated above. At least, it's what it looks like/

**Comment by qnebra** [comment]
I have this same issue on lmms 1.1.3 from kxstudio in vivid.

**Comment by tresf** [comment]
@badosu is this a valid bug with our code, or a problem with the way these downstream packages are created?

**Comment by badosu** [comment]
@tresf I'd have to try a previous release to test this, I really don't know. Gonna take a look.

**Comment by Astrognome** [comment]
I can also produce this issue on Arch using the provided packages. This also happens when I use the git version from AUR.

**Comment by Gnurou** [comment]
Same issue, also with Arch Linux 64 bit. Will try to investigate a bit further if I can find the time.

**Comment by Umcaruje** [comment]
Can any of you confirm this bug on Arch if you manually compile LMMS and not use the packages?

**Comment by obszczymucha** [comment]
I've built multiple versions from the source. Same problem. I'll try and build the latest 1.1.3 version and let you know the outcome.

**Comment by obszczymucha** [comment]
Just built 1.1.3 - the issue is still there.

**Comment by Gnurou** [comment]
Yes, this is from a source build. I have started looking at it, but could not rootcause it yet, excepted that it seems to be a miscommunication between the main process and RemoteVstPlugin. That part of the code is not particularly easy to comprehend to new eyes. Here is the top of the backtrace from the main process:

    #0  0x00007ffff7bcdc1d in recvmsg () from /usr/lib/libpthread.so.0
    #1  0x00007ffff082e917 in ?? () from /usr/lib/libxcb.so.1
    #2  0x00007ffff082f188 in ?? () from /usr/lib/libxcb.so.1
    #3  0x00007ffff28abdc8 in ?? () from /usr/lib/libX11.so.6
    #4  0x00007ffff28abf2b in ?? () from /usr/lib/libX11.so.6
    #5  0x00007ffff28ac23d in _XEventsQueued () from /usr/lib/libX11.so.6
    #6  0x00007ffff289db10 in XEventsQueued () from /usr/lib/libX11.so.6
    #7  0x00007ffff713bbf7 in ?? () from /usr/lib/libQtGui.so.4
    #8  0x00007ffff3fcb1bd in g_main_context_prepare () from /usr/lib/libglib-2.0.so.0
    #9  0x00007ffff3fcbba8 in ?? () from /usr/lib/libglib-2.0.so.0
    #10 0x00007ffff3fcbd8c in g_main_context_iteration () from /usr/lib/libglib-2.0.so.0
    #11 0x00007ffff6952044 in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) ()
   from /usr/lib/libQtCore.so.4
    #12 0x00007ffff713c156 in ?? () from /usr/lib/libQtGui.so.4
    #13 0x00007ffff69255ce in QCoreApplication::processEvents(QFlags<QEventLoop::ProcessEventsFlag>, int) ()
   from /usr/lib/libQtCore.so.4
    #14 0x000000000053607e in RemotePluginBase::waitForMessage (this=0x1b6ae00, _wm=..., _busy_waiting=true)
    at /home/gnurou/Projects/lmms/lmms/include/RemotePlugin.h:954
    #15 0x0000000000537fc7 in RemotePlugin::waitForInitDone (this=0x1b6ae00, _busyWaiting=true)
    at /home/gnurou/Projects/lmms/lmms/include/RemotePlugin.h:690
    #16 0x00007fffa2732d19 in VstPlugin::tryLoad (this=0x1b6add0, remoteVstPluginExecutable=...)
    at /home/gnurou/Projects/lmms/lmms/plugins/vst_base/VstPlugin.cpp:183
    #17 0x00007fffa27327dc in VstPlugin::VstPlugin (this=0x1b6add0, _plugin=...)
    at /home/gnurou/Projects/lmms/lmms/plugins/vst_base/VstPlugin.cpp:102
    #18 0x00007fffa1bf4f51 in vestigeInstrument::loadFile (this=0x7fffeafb7da0, _file=...)
    at /home/gnurou/Projects/lmms/lmms/plugins/vestige/vestige.cpp:260
    #19 0x00007fffa1bf73f8 in VestigeInstrumentView::openPlugin (this=0x18d8440)
    at /home/gnurou/Projects/lmms/lmms/plugins/vestige/vestige.cpp:650
    #20 0x00007fffa1bfd178 in VestigeInstrumentView::qt_static_metacall (_o=0x18d8440, 
    _c=QMetaObject::InvokeMetaMethod, _id=1, _a=0x7fffffffd060)
    at /home/gnurou/Projects/lmms/lmms/build/plugins/vestige/moc_vestige.cxx:217

What is weird is that it seems to happen only on Arch Linux?

Let me know if I can provide more information.

**Comment by hydrix9** [comment]
confirming from arch x64 with latest versions of wine and lmms

**Comment by Narfinger** [comment]
I think I tracked down the wait to [RemotePlugin.h:293](https://github.com/LMMS/lmms/blob/stable-1.1/include/RemotePlugin.h#L293). We are seeming to wait on a semaphore.

Interesting enough, I tried to change to define USE_QT_SEMAPHORES which still crashes lmms in the same way but it gives me the GUI of the VST.

**Comment by tresf** [comment]
I'd love to reproduce this problem and help troubleshoot, but I simply don't have the patience for a 50 page wiki article to get the OS loaded in a VM and I think this is an unreasonable request for the development team of music software.

Can someone provide a working VM image?  By "working", I mean something I can boot and use that is pre-partitioned, has networking installed, a boot loaded, et. al?

I've read a few quickstart guides, and even they seem to take hours to complete.

-Tres

**Comment by Wallacoloo** [comment]
@tresf you could try one of the Arch-based [distros](https://wiki.archlinux.org/index.php/Arch_based_distributions_%28active%29).

And if you're running it in a VM you shouldn't need to do any boot loader stuff.

**Comment by tresf** [comment]
> @tresf you could try one of the Arch-based distros.

Tremedous. I haven't heard of or tried any of those.  Any recommendation?

**Comment by tresf** [comment]
Manjaro Linux seems to be #\8 on distrowatch.  I'll check that one out.

**Comment by Narfinger** [comment]
Another thing which I noticed in a debug session is that RemotePluginBase::waitForMessage is busy waiting while zero messages are left. Perhaps isInvalid() is not updated correctly?

**Comment by tresf** [comment]
Sorry I hadn't yet posted an update on this...  My Manjaro install experiences the exact symptoms of that above.

I think the long term solution is to add better exception handling to our `RemotePlugin` code.  When LMMS talks to an external process (this happens for ZynAddSubFX, VST(fx), VST(i)) we have scenarios where our shared memory logic fails with no sign of why.

I believe this is in part due to the choice to use Qt Shared memory for some platforms, but not all.  I assume shared memory started as a *nix-Only feature and matured into using Qt's more platform independent implementation.

Here's an article that explains a bit how we can do better debugging:  http://stackoverflow.com/a/7497455/3196753

Although I'm curious if we should just switch our code to use Qt for all platforms, since it comes as a dependency on our codebase anyway.

The only person on the project I've ever seen actually understand this is the author, @tobydox.  Perhaps we'll get lucky and get some insight.

**Comment by mkondel** [comment]
The shmget error goes away after installing wine1.7, at least on ubuntu.  The VSTs still do not load, even after compiling the latest master.  It's not only Arch where this happens.

**Comment by tresf** [comment]
I believe this is a duplicate of [`forums/#1640`](https://lmms.io/forum/viewtopic.php?t=1640) and possibly #2196.

FYI, Here's Toby's response from January 2014 on the topic... 

> Tobias Doerffel 2014-01-23 09:52:53 CST
>
> Hi,
>
> as the original main developer of LMMS and its VST support layer, I'm joining. First of all thanks for your efforts to fix this issue! Regarding your questions:
> 
> The source code of the application which is launched externally using WINE can be found here:
>
> https://github.com/LMMS/lmms/blob/stable-0.4/plugins/vst_base/RemoteVstPlugin.cpp
>
> At line 692 you can find the code which retrieves the X11 window ID. This code always has been working. Is there any other recommended method?
>
>
> Inside LMMS we use QX11EmbedContainer to embed the window inside our MDI sub window:
>
> https://github.com/LMMS/lmms/blob/stable-0.4/plugins/vst_base/VstPlugin.cpp
>
> at line 245 ff. So basically we don't do any magic here.

-Tres

**Comment by tresf** [comment]
Some more background information:

https://github.com/LMMS/lmms/issues/1468#issuecomment-71485618

**Comment by mkondel** [comment]
I get a message like this on the console when VST is stuck in the infinite loop:
```
fixme:ole:RemUnknown_QueryInterface No interface for iid {00000019-0000-0000-c000-000000000046}
```
Searching on google shows lots of Wine related issues, but the fixes are all over the place, these cant help with VST...  Ex:
```
- After the latest update [of Wine prolly], my tax return software works
- installing internet explorer in the same wineprefix solved it for me
- the problem was that I didn't have lib32-nvidia-libgl
```

**Comment by tresf** [comment]
Ok, I found something that wasn't in any of our forums or bug reports.... An upstream bug that was filed with Wine and claims to be fixed with 1.7.11.  Not sure if this is related, but it has the same copy/paste message that came from Toby in the forums... 

https://bugs.winehq.org/show_bug.cgi?id=35347
https://bugreports.qt.io/browse/QTBUG-36446

Wine claims to have patched this in version 1.7.11 (My Manjaro install is on 1.7.49), but there is a recommendation in there which we may find useful as it describes that if Qt eventually fixes this, our code can cause issues.

> If the problems with QT are fixed later you will probably experience an additional problem, which is caused by the fact that you make the window visible before its embedded. This never works that well, and I would recommend to do this after the reparenting step (by using some additional ipc messages).

Sorry if this is all inaccurate such as in the event this bug is actually fixed.  Trying to dig up more info on this... 

-Tres

**Comment by mkondel** [comment]
If RemoteVstPlugin.cpp is the app that runs in Wine, would that be a good starting point to debug what's breaking?  I seem to get it to printf to the console, at least getting to line 357 at the moment.

**Comment by tresf** [comment]
> If RemoteVstPlugin.cpp is the app that runs in Wine, would that be a good starting point to debug what's breaking?

Sure, just take note that `RemotePlugin` code is getting executed here as well.

**Comment by mkondel** [comment]
The execution seems to hang here:
https://github.com/LMMS/lmms/blob/master/plugins/vst_base/RemoteVstPlugin.cpp#L407
Line 407 gets called, but the following code, in RemotePlugin, is never run:
https://github.com/LMMS/lmms/blob/master/include/RemotePlugin.h#L922

**Comment by Narfinger** [comment]
I removed the non-qt ShmMem and Semaphore code. It can be found in master of my repository https://github.com/Narfinger/lmms.

However, now I get a wine backtrace which I do not know how to debug.

**Comment by tresf** [comment]
@Narfinger do to this proper, should you tackle `RemotePlugin` as well?

With your branch, I'm getting this on Ubuntu 12.04.  I'm guessing because some of the dependencies aren't in order for non-Windows `QSharedMemory` usage.

```
[ 81%] Generating RemoteVstPlugin
In file included from /home/tres/lmms/plugins/vst_base/RemoteVstPlugin.cpp:35:0:
/home/tres/lmms/include/RemotePlugin.h:41:27: fatal error: QtCore/QtGlobal: No such file or directory
compilation terminated.
winegcc: g++ failed
make[2]: *** [plugins/vst_base/RemoteVstPlugin] Error 2
make[1]: *** [plugins/vst_base/CMakeFiles/vstbase.dir/all] Error 2
make: *** [all] Error 2
tres@ubuntu-1204:~/lmms/build$ vi /home/tres/lmms/plugins/vst_base/RemoteVstPlugin.cpp
tres@ubuntu-1204:~/lmms/build$ 
```


-Tres

**Comment by Narfinger** [comment]
Hey.
I just updated my branch, indeed I forgot something. I also had to change some cmake file. This was rather ad-hoc.

But good news. Synth1 loads for me. If it loads for other people (somebody should probably test this on windows) I can cleanup the changes and do a pull request.

**Comment by mkondel** [comment]
Any quick ideas? I tried to build your fork on master:
```
Scanning dependencies of target vibedstrings
[ 44%] Building CXX object plugins/vibed/CMakeFiles/vibedstrings.dir/vibed.cpp.o
/usr/bin/ld: cannot find -lQtCore
collect2: error: ld returned 1 exit status
winegcc: g++ failed
plugins/vst_base/CMakeFiles/vstbase.dir/build.make:53: recipe for target 'plugins/vst_base/RemoteVstPlugin' failed
make[2]: *** [plugins/vst_base/RemoteVstPlugin] Error 2
CMakeFiles/Makefile2:8539: recipe for target 'plugins/vst_base/CMakeFiles/vstbase.dir/all' failed
make[1]: *** [plugins/vst_base/CMakeFiles/vstbase.dir/all] Error 2

```

**Comment by tresf** [comment]
> If it loads for other people (somebody should probably test this on windows) I can cleanup the changes and do a pull request.

A pull request will do the Windows testing for you automatically due to our Travis-CI integration. :+1:

**Comment by tresf** [comment]
> Any quick ideas? I tried to build your fork on master:

I'm not having any luck from Ubuntu side either... A pull request will show the errors for Ubuntu in Travis-CI too but in my case it might be a 32/64bit lib thing.  I'm running a clean build now to capture the exact error.

**Comment by tresf** [comment]
Here's my error (Ubuntu 12.04) which is the same thing @mkondel is receiving, just more verbose.

Also @mkondel don't use `-j2` or `-j4` when you're posting a debug message as you'll get success messages mixed into error messages, as are your vibed lines.

At a glance, I think the `-m32` switch wants the `32-bit` libpthread.so perhaps?  Sorry if I'm of no assistance. 

```
[ 81%] Generating RemoteVstPlugin
/usr/bin/ld: skipping incompatible /usr/lib/x86_64-linux-gnu/libpthread.so when searching for -lpthread
/usr/bin/ld: skipping incompatible /usr/lib/x86_64-linux-gnu/libpthread.a when searching for -lpthread
/usr/bin/ld: skipping incompatible /usr/lib/x86_64-linux-gnu/libQtCore.so when searching for -lQtCore
/usr/bin/ld: cannot find -lQtCore
/usr/bin/ld: skipping incompatible /usr/lib/x86_64-linux-gnu/libwine.so when searching for -lwine
/usr/bin/ld: skipping incompatible /usr/lib/x86_64-linux-gnu/libm.so when searching for -lm
/usr/bin/ld: skipping incompatible /usr/lib/x86_64-linux-gnu/libm.a when searching for -lm
/usr/bin/ld: skipping incompatible /usr/lib/x86_64-linux-gnu/libc.so when searching for -lc
/usr/bin/ld: skipping incompatible /usr/lib/x86_64-linux-gnu/libc.a when searching for -lc
```

**Comment by Narfinger** [comment]
This happens because I hacked the Cmake file to work for my system but did not check how to do this in general because I didn't have time. It should be fixed in the pull request I will do later.

**Comment by ichiman94** [comment]
@Narfinger

I have ArchLinux 64bit too, and your fork with the latest commit allows me to run 32bit Synth1 successfully, however the 64bit version stalls lmms.

(Oh BTW, to compile lmms, install gcc-multilib and lib32-qt4 packages)

**Comment by Narfinger** [comment]
I don't know about 64 bit, did 64 bit plugins ever work? It doesn't look like it.

For the bugs above: I think @ichiman94 is right, you need the 32bit qt4 packages (at least QtCore). I have a fixed-vestige branch on my repository, however there is a slight problem:
At the moment the -lQtCore flag is hardcoded in the cmake file as find_package in cmake will just find the 64 bit thing but we need to explicitly link against the 32 bit libraries.

I at the moment do not know how to have a good cmake file. Somebody from the cmake channel said we should look into ExternalProject and make this a 32 bit project (because we can only make a whole cmake project into 32bit). But I need to read up on this.

**Comment by tresf** [comment]
> At the moment the -lQtCore flag is hardcoded in the cmake file as find_package in cmake will just find the 64 bit thing but we need to explicitly link against the 32 bit libraries.

According to this mailing list archive `-m32` should be enough.  http://www.cmake.org/pipermail/cmake/2013-February/053595.html

Although I have feeling these paths are borrowed from `qmake` rather than `cmake` so some other type of trickery may be needed although this posts suggests it might just be a missing flag... 

http://stackoverflow.com/a/19310455/3196753

**Comment by Zeioth** [comment]
I'm experiencing exactly the same problem under  Xubuntu 15.04 x64 + wine-1.7.50
I'm glad to hear the solution it's on the way, I'll keep in the loop.

More system info -> https://i.imgur.com/BLCu4wy.png

**Comment by Wallacoloo** [comment]
@tresf for future testing on arch (rather than a derivative like Manjaro), I came across an unofficial arch installer that you may find useful: https://sourceforge.net/projects/dh-arch-installer/

**Comment by tresf** [comment]
> `Install arch-linux in a damn hurry`

lol. 

@Wallacoloo is this recommended over Manjaro?  I don't mind either approach, BTW.

**Comment by Wallacoloo** [comment]
@tresf you'll be using identical base system, repositories and packages as
a normal arch system (I would imagine Manjaro applies some branding and/or
patching to the normal arch packages), which means you should have more
success in replicating any arch-only bugs.

But if you already have Manjaro up and running, then no need to try this
method unless you ever fail to reproduce an arch-only bug.

**Comment by tresf** [comment]
:+1:

**Comment by thebeav26** [comment]
I am experiencing this issue still on Arch 64. According to this (https://github.com/LMMS/lmms/pull/2304) should the issue now be fixed in the git version of lmms? I tried building the git version last night (9Sep2015) and still had the same issue. Not sure if I had the lib32-qt4 package installed or not at the time, though.

Can someone please help with a workaround suggestion so I can use VST's?

**Comment by tresf** [comment]
Although #2304 potentially fixes this issue, it was never merged into our source code due to it breaking backwards compatibility for the old VST behavior. 

If you clone the branch from the #2304 fork, you will have working VSTs.  The author of that commit needs to make some changes before being accepted into our git source tree.

**Comment by thebeav26** [comment]
@tresf So I cloned @Narfinger 's branch and I'm still unable to load up Synth1. However, with the new version, it at least doesn't  freeze the program in an infinite loop. You can continue using LMMS just fine after. In the terminal window I'm running LMMS from, I get the error message "remote plugin died! invalidating now." and then LMMS just goes on its way without loading the VST.

**Comment by tresf** [comment]
I assume you are using the `fix-vestige` branch?  Perhaps he can help getting it running then...

**Comment by thebeav26** [comment]
Yes, I forked the main LMMS repo and then merged it with his fix-vestige branch.

**Comment by ichiman94** [comment]
Are you trying to load the 64bit Synth1? Try out with the 32bit version of Synth1.

**Comment by thebeav26** [comment]
@ichiman94 I tried both the 32 bit and the 64 bit *.dll's for Synth1 and a few other 32 bit VST's as well actually.

**Comment by thebeav26** [comment]
@Narfinger any ideas?

**Comment by Narfinger** [comment]
Sorry I do not have any idea why this does not work.

I got stuck trying to split the QSemaphore and QSHM code into separate classes because the shm code needs a bit more understanding and I did not have time yet.

Related to that I also tried using boost instead of Qt/linux native semaphores and queue. However, just using boost semaphore already gives me weird compile errors. Does anybody know if winegcc and boost are just incompatible? I really did not find much information on the net.

**Comment by Wallacoloo** [comment]
Pretty sure boost would be compatible with mingw (winegcc?), especially as
its intention is to be cross-platform.

But I would be hesitant in adding boost as another dependency to lmms, on
top of the already huge Qt. If you're just using it for semaphores, I'm
pretty sure the c++11 std library includes those anyway.

**Comment by Narfinger** [comment]
The advantage with boost is that there is no linking dependency that is why I looked into it. Additionally, I was thinking that it might actually remove the FIFO implemented in the class entirely and replace it by a standard boost class that is thread safe. I will check if there is something in the c++11 std library for semaphores and shared memory.

**Comment by zommuter** [comment]
Same issue here - just for the fun of it ran the Windows (64bit) version of LMMS via wine, where this works just fine...

**Comment by kernelcrasher** [comment]
I see that the split patch in #2390 has passed all checks. Does this mean merge and possibly VSTs loading correctly soon?

**Comment by Narfinger** [comment]
You are welcome to try but I wouldn't merge it into main yet. I still need to look at the problem posted by @tresf. I hope I can find time next week to check it but I cannot guarantee anything.

**Comment by kernelcrasher** [comment]
Thank you, and thanks a lot for the work.

**Comment by tresf** [comment]
> I see that the split patch in #2390 has passed all checks. Does this mean merge and possibly VSTs loading correctly soon?

The mentioned PR doesn't actually address the VST bug yet per https://github.com/LMMS/lmms/pull/2390#issuecomment-148114571

So even when the Ubuntu 12.04 crash is fixed and the code split is merged, the option to use `QSharedMemory` (as opposed to the native IPC logic which is broken on some platforms) per this comment (unless I'm missing something):

> I did not yet implement the switching to qt part.

**Comment by kernelcrasher** [comment]
From reading the linked issues, WineHQ and the Qt mailing list, it appears that the problem has its roots in the Qt implementation of a function called sendXEmbedMessage(). The arguments are actually passed in the wrong order, which causes everything to break.
See:
https://bugs.winehq.org/show_bug.cgi?id=35347
The issue was reported on the Qt bug tracker in 2014 and has remained unsolved **to this day**.
(https://bugreports.qt.io/browse/QTBUG-36446)

Seeing as the Linux users are stuck with a bug that makes most virtual instruments unusable with an estimated time to fix of *never*, is it possible to figure out some workaround?

**Comment by Wallacoloo** [comment]
Given the name of the function ("X"), perhaps the bug can be avoided by
running LMMS under wayland instead of X11?

**Comment by tresf** [comment]
@Wallacoloo perhaps, but if this is a Qt bug, I'm not sure how it broke yet again.  Perhaps someone reverted the Wine code accidentally in a cleanup effort.  I see mentions of `QX11EmbedWidget` going away with Qt5, so perhaps that's a more viable option?

**Comment by CtrlFreak1337** [comment]
So then, installing Qt5 and following the build instructions on the wiki might solve the problem? I'll give that a try and see what works.

**Comment by amagnolo** [comment]
This affects also Lubuntu 15.10 x64 + wine-1.8 + LMMS 1.1.3 from kxstudio repositories.
The tested VSTs (Synth1, Crystal) work all right when loaded in Carla instead of LMMS.

**Comment by kamnxt** [comment]
Affects LMMS 1.1.90 built from source with wine-1.9.6 on Arch Linux x64.

Tried both 32 and 64 bit VSTs.

**Comment by geldedus** [comment]
Affects LMMS 1.1.3 (Linux/x86_64, Qt4.8.6, GCC 4.8.2) on Ubuntu 15.10

**Comment by kernelcrasher** [comment]
It has been over a a year now.
I was wondering if there could be some temporary workaround related to the Qt bug.
Wouldn't it be possible to intentionally pass the arguments in the wrong order so that sendXEmbedMessage actually passes them in the right order?
Also, it seems like someone on WineHQ has produced a temporary workaround:
http://source.winehq.org/patches/data/101921
has this been tested?

**Comment by tresf** [comment]
> has this [workaround] been tested?

Assuming it's the same bug, doesn't that require a patch to X11?

> I was wondering if there could be some temporary workaround related to the Qt bug.

There is, @Narfinger wrote one https://github.com/LMMS/lmms/pull/2304.

Unfortunately, this was closed as the original author of the VST bridge didn't like the idea of pulling IPC support.  That spawned the first step to split the code https://github.com/LMMS/lmms/pull/2390 whereas the next step is to allow the Qt method to be toggled on.

The first step had a bug which is explained the PR.   No further progress has been made, but please feel free to hack away at it.

**Comment by amagnolo** [comment]
I'm sorry I don't understand, is there currently any way to use VSTs with LMMS in Linux? Or this feature is broken and was abandoned one year ago?
What's the workaround to this bug, should I change Linux distribution, use an alternative to VeSTige, ...? Please don't tell me to install LMMS in Windows. ;-)

**Comment by tresf** [comment]
> I'm sorry I don't understand, is there currently any way to use VSTs with LMMS in Linux? Or this feature is broken and was abandoned one year ago?

Unless using Carla, not that I'm aware of, at least not on most modern Linux versions (VSTs load fine on Ubuntu 12.04 for me).

> What's the workaround to this bug

You could always clone and build the branch in #2304 and try that out, but of course that's not up to date with the stability changes we've had to master.

> should I change Linux distribution, use an alternative to VeSTige

Or an alternative to LMMS, if it's a deal breaker for your needs.  We need dev help on this.

> Please don't tell me to install LMMS in Windows. ;-)

:)

**Comment by tresf** [comment]
Can someone please test this patch on Arch #2728

I'm testing on ElementaryOS Freya and it seems to be working so as long as I putting the linking back to the old method.

![screen shot 2016-04-14 at 7 19 37 pm](https://cloud.githubusercontent.com/assets/6345473/14546810/e9dfd88a-0275-11e6-98bf-08fdfcb3bef2.png)

**Comment by liushuyu** [comment]
I also have the lastest Arch Linux, and I can test this out very soon.

**Comment by liushuyu** [comment]
No, unfortunately, still didn't work

**Comment by kernelcrasher** [comment]
Anyone tried the patch?

**Comment by liushuyu** [comment]
>Anyone tried the patch?

I have tried this... No it didn't work...

**Comment by kernelcrasher** [comment]
The Qt bug report has been updated (https://bugreports.qt.io/browse/QTBUG-36446): apparently, they won't patch Qt 4 and Qt 5 doesn't have the functionality that allows for writing host applications.
So, either this is fixed on the LMMS side by adopting some other way of using VSTs, or they will remain permanently unusable.

**Comment by jasp00** [comment]
According to https://github.com/LMMS/lmms/pull/2739#issuecomment-220817059, this issue is fixed.

**Comment by liushuyu** [comment]
According to #2739, this issue is fixed.

**Comment by keantoken** [comment]
I have this problem on Debian Sid, but only with a 64-bit VST.

RemotePluginClient::shmget: No such file or directory
RemoteVstPlugin.cpp::shmget: No such file or directory
RemoteVstPlugin.cpp: Failed to initialize shared memory for VST synchronization.
 (VST-host synchronization will be disabled)

The VST fails to load, and then 'wineserver64 -p0' turns into a zombie and eats my CPU. I am using VSTSynthfont64. The 32-bit version of this synthfont does have this problem.
